tests/test_funcs.py

```diff
@@ -271,7 +272,7 @@ def test_simple(self):
         assert_allclose(w, xp.flip(w, axis=0))
 
     @pytest.mark.parametrize("x", [0, 1 + 3j])
-    def test_dtype(self, x):
+    def test_dtype(self, x: int | complex):
```

Comment:
Perhaps you're already aware of this; but in the static typing word, this is equivalent to `complex` (even though it isn't at runtime...). But you could also argue that it's better to be explicit about it, and type it as `int | complex` instead of `complex` like you've currently done.

Reply:
This little issue is going to be something you're going to have to decide how to deal with in typing the array API. 

In the array API (and the numerics world in general), integers and floats are *not* compatible types. The array API specifically discourages implicitly casting from one to the other. But according to [Python typing](https://peps.python.org/pep-0484/#the-numeric-tower), any function that accepts floats should also accept integers. It's going to be even more of an issue now that we are adding support for Python scalars to functional APIs (https://github.com/data-apis/array-api/issues/807). There's plenty of functions that accept float but are not required to accept int (and even some that *must not* accept int). See also https://github.com/data-apis/array-api/pull/862#discussion_r1857364664.

Reply:
I've ran into this issue more times than I can count...
Recently I found a workaround for (at least) `int` that works with both mypy and pyright, which is available in [`optype`](https://github.com/jorenham/optype) as `optype.typing.JustInt` ([docs](https://github.com/jorenham/optype/blob/master/README.md#just-types)).
There's also the more general `optype.typing.Just`, but because of an issue in typeshed (which recently has been fixed), and a bug in mypy, this currently doesn't work for `float` and `complex`. If the next pyright release bundles the latest typeshed, then it should at least be possible in pyright to use `optype.typing.Just[float]` so that it rejects `int` and `bool` and "just" accepts `float` instances.

Reply:
i'm planning to address this situation entirely in the basedtyping ecosystem, perhaps not so relevant to the 484 ecosystem

Reply:
Is it really an issue? The `int |` was kept in a number of places for clarity, but we had the discussion between about `int | float` and `float` meaning the same in type annotation.

Note that it's _very_ different between array library dtypes and Python scalars. For Python scalars the considerations that made us avoid mixed int-float promotion rules all just don't exist.

Reply:
Yea it's a pretty big issue, especially for numpy, see https://github.com/numpy/numpy/issues/27032

Reply:
But apart from the issue of the overlapping overloads, the worst thing is that it's type-unsafe. For example

```py
def get_hex(x: float, /) -> str:
    return x.hex()

get_hex(42)  # accepted
```

which at runtime results in

```
AttributeError: 'int' object has no attribute 'hex'
```